restcomm/restcomm.interpreter/src/main/java/org/restcomm/connect/interpreter/NumberSelectorService.java

```diff
@@ -0,0 +1,245 @@
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Adding Plus Num:" + plusNum);
+                }
+                numberQueries.add(plusNum);
+            }
+        }
+        return numberQueries;
+    }
+
+    private IncomingPhoneNumber findSingleNumber(String number,
+            Sid sourceOrganizationSid, Sid destinationOrganizationSid) {
+        IncomingPhoneNumber matchedNumber = null;
+        IncomingPhoneNumberFilter.Builder filterBuilder = IncomingPhoneNumberFilter.Builder.builder();
+        filterBuilder.byPhoneNumber(number);
+        if (destinationOrganizationSid != null) {
+            filterBuilder.byOrgSid(destinationOrganizationSid.toString());
+        }
+        if (sourceOrganizationSid!= null &&
+                !sourceOrganizationSid.equals(destinationOrganizationSid)) {
+            filterBuilder.byPureSIP(Boolean.FALSE);
+        }
+        IncomingPhoneNumberFilter numFilter = filterBuilder.build();
+        if (logger.isDebugEnabled()) {
+            logger.debug("Searching with filter:" + numFilter);
+        }
+        List<IncomingPhoneNumber> matchedNumbers = numbersDao.getIncomingPhoneNumbersByFilter(numFilter);
+        if (logger.isDebugEnabled()) {
+            logger.debug("Num of results:" + matchedNumbers.size());
+        }
+        //we expect a perfect match, so first result taken
+        if (matchedNumbers != null && matchedNumbers.size() > 0) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Matched number with filter:" + matchedNumbers.get(0));
+            }
+            matchedNumber = matchedNumbers.get(0);
+        }
+        return matchedNumber;
+    }
+
+    private IncomingPhoneNumber findByNumber(List<String> numberQueries,
+            Sid sourceOrganizationSid, Sid destinationOrganizationSid) {
+        IncomingPhoneNumber matchedNumber = null;
+
+        int i = 0;
+        while (matchedNumber == null && i < numberQueries.size()) {
+            matchedNumber = findSingleNumber(numberQueries.get(i),
+                    sourceOrganizationSid, destinationOrganizationSid);
+            i = i + 1;
+        }
+        return matchedNumber;
+    }
+
+    public IncomingPhoneNumber searchNumber(String phone,
+            Sid sourceOrganizationSid, Sid destinationOrganizationSid) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("getMostOptimalIncomingPhoneNumber: " + phone
+                    + ",srcOrg:" + sourceOrganizationSid
+                    + ",destOrg:" + destinationOrganizationSid);
+        }
+        List<String> numberQueries = createPhoneQuery(phone);
+
+        IncomingPhoneNumber numberfound = findByNumber(numberQueries, sourceOrganizationSid, destinationOrganizationSid);
+        if (numberfound == null) {
+            //if no perfect match found, try with special star number
+            numberfound = findSingleNumber("*", sourceOrganizationSid, destinationOrganizationSid);
```

Comment:
@jaimecasero I think REGEX numbers should have higher priority than the `*` which should be the last choice if nothing else matched.
The order should be from the most specific to the catch-em-all option, so:
specific number -> regex -> '*'

Reply:
ummm,interesting, i think this coming from old code, let me check what is there previously. The requirement is not clear to me, but your suggestion makes sense to me..

Reply:
@gvagenas  you are right, special star matching should be last attempt,even after regexes, i will change impl/test for this